contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+
+## Background
+
+ThirdPartyResource (TPR) is deprecated and CustomResourceDefinition (CRD) is the successor which solves the fundamental [issues](https://github.com/kubernetes/features/issues/95) of TPRs to form a stable base for further features.
+
+Currently we do not provide validation for CustomResources (CR), i.e. the CR payload is free-form JSON. However, one of the most requested [[1](https://github.com/kubernetes/features/issues/95#issuecomment-296416969)][[2](https://github.com/kubernetes/features/issues/95#issuecomment-298791881)] features is validation and this proposal seeks to add it.
+
+## Goals
+
+1. To provide validation for CustomResources using a declarative specification language for JSON data.
+2. To keep open the door to add other validation mechanisms later.<sup id="f2">[2](#footnote2)</sup>
+3. To allow server-side validation.
+4. To be able to integrate into the existing client-side validation of kubectl.
+5. To be able to define defaults in the specification (at least in a follow-up after basic validation support).
+
+## Non-Goals
+
+1. The JSON-Schema specs can be used for creating OpenAPI documentation for CRs. The format is compatible but we won’t propose an implementation for that.
+2. A turing-complete specification language is not proposed. Instead a declarative way is proposed to express the vast majority of validations.
+3. For now, CRD only allows 1 version at a time. Supporting multiple versions of CRD and/or conversion of CRD is not within the scope of this proposal.
+
+## Proposed Extension of CustomResourceDefinition
+
+We propose to add a field `validation` to the spec of a CustomResourceDefinition. As a first validation format we propose to use [JSON-Schema](http://json-schema.org/) under `CRD.Spec.Validation.JSONSchema`.
+
+JSON-Schema is a [standardized](https://tools.ietf.org/html/draft-zyp-json-schema-04) declarative specification language. Different keywords may be utilized to put constraints on the data. Thus it provides ways to make assertions about what a valid document must look like. 
+
+It is already used in Swagger/OpenAPI specs in Kubernetes and hence such a CRD specification integrates cleanly into the existing infrastructure of the API server which serves these specifications,
+* into kubectl which is able to verify YAML and JSON objects against the returned specification.
+* With the https://github.com/go-openapi/validate library, we have a powerful JSON-Schema validator which can be used client and server-side.
+
+## API Types
+
+The schema is referenced in [`CustomResourceDefinitionSpec`](https://github.com/kubernetes/kubernetes/commit/0304ef60a210758ab4ac43a468f8a5e19f39ff5a#diff-0e64a9ef2cf809a2a611b16fd44d22f8). `Validation` is of the type `CustomResourceValidation`. The JSON-Schema is stored in a field of `Validation`. This way we can make the validation generic and add other validations in the future as well.
```

Comment:
To address @brendanburns's comment https://github.com/kubernetes/community/pull/708#discussion_r122353129:
```
This way we can make the validation generic and add other validation options in the future. Even multiple validation options could be used at the same time within one CustomResourceDefinition.
```